include/executor/engine/atomic.ipp

```diff
@@ -0,0 +1,330 @@
+#include "executor/executor.h"
+#include "runtime/instance/memory.h"
+
+#include <cstdint>
+
+namespace WasmEdge {
+namespace Executor {
+
+template <typename T>
+TypeT<T> Executor::runAtomicWaitOp(Runtime::StackManager &StackMgr,
+                                   Runtime::Instance::MemoryInstance &MemInst,
+                                   const AST::Instruction &Instr) {
+
+  const uint32_t BitWidth = sizeof(T) * 8;
+  ValVariant Address = StackMgr.pop();
+  ValVariant Val = StackMgr.pop();
+  [[maybe_unused]] ValVariant Timeout = StackMgr.pop();
+
+  StackMgr.push(Address);
+  runLoadOp<T>(StackMgr, MemInst, Instr, BitWidth);
+  ValVariant &Loaded = StackMgr.getTop();
+
+  if (Loaded.get<T>() == Val.get<T>()) {
+    StackMgr.push(ValVariant(0));
+  } else {
+    StackMgr.push(ValVariant(1));
+  }
+  // Wait will be supported in C++20
+  // TODO: Implement Timeout
+  return {};
+}
+
+template <typename T, typename I>
+TypeT<T> Executor::runAtomicLoadOp(Runtime::StackManager &StackMgr,
+                                   Runtime::Instance::MemoryInstance &MemInst,
+                                   const AST::Instruction &Instr) {
+
+  const uint32_t BitWidth = sizeof(I) * 8;
+  ValVariant RawAddress = StackMgr.pop();
+  int32_t Address = RawAddress.get<int32_t>();
+  if ((Address & ((BitWidth >> 3U) - 1)) != 0) {
+    spdlog::error(ErrCode::UnalignedAtomicAccess);
+    spdlog::error(
+        ErrInfo::InfoInstruction(Instr.getOpCode(), Instr.getOffset()));
+    return Unexpect(ErrCode::UnalignedAtomicAccess);
+  }
+
+  // make sure the address no OOB with size I
+  I *RawPointer = MemInst.getPointer<I *>(Address);
```

Comment:
Do you consider to implement the atomic load and atomic store in the memory instance class? Such as the `loadValue()` function.